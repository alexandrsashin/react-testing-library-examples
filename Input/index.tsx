import { ChangeEvent } from "react";

interface InputProps {
  name: string;
  value?: string;
  type?: "text" | "number";
  onChange: (event: {
    name: string;
    value: string;
    valueAsNumber: number;
  }) => void;
}

const Input = ({ name, value, type = "text", onChange }: InputProps) => {
  return (
    <input
      name={name}
      value={value}
      type={type}
      onChange={({ target }: ChangeEvent<HTMLInputElement>) =>
        onChange({
          name: target.name,
          value: target.value,
          valueAsNumber: target.valueAsNumber,
        })
      }
    />
  );
};

export default Input;
