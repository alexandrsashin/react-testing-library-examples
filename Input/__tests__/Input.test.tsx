import {
  cleanup,
  fireEvent,
  screen,
  render,
  RenderResult,
} from "@testing-library/react";
import Input from "..";

const onChange = jest.fn();

describe("<Input />", () => {
  let view: RenderResult;

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  describe("with type=text", () => {
    beforeEach(() => {
      view = render(
        <Input name="name" value={undefined} type="text" onChange={onChange} />
      );
    });

    it("should render", () => {
      expect(view.asFragment()).toMatchSnapshot();
    });

    it("onChange() is called after input change", () => {
      const input = screen.getByRole("textbox");

      fireEvent.change(input, { target: { value: "test text" } });
      expect(onChange).toHaveBeenCalledWith({
        name: "name",
        value: "test text",
        valueAsNumber: NaN,
      });
    });
  });

  describe("with type=number", () => {
    beforeEach(() => {
      view = render(
        <Input
          name="name"
          value={undefined}
          type="number"
          onChange={onChange}
        />
      );
    });

    it("should render", () => {
      expect(view.asFragment()).toMatchSnapshot();
    });

    it("onChange() is called after input change", () => {
      const input = screen.getByRole("spinbutton");

      fireEvent.change(input, { target: { value: "4" } });
      expect(onChange).toHaveBeenCalledWith({
        name: "name",
        value: "4",
        valueAsNumber: 4,
      });
    });
  });
});
