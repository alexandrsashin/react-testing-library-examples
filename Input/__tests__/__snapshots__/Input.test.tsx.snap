// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> with type=number should render 1`] = `
<DocumentFragment>
  <input
    name="name"
    type="number"
    value=""
  />
</DocumentFragment>
`;

exports[`<Input /> with type=text should render 1`] = `
<DocumentFragment>
  <input
    name="name"
    type="text"
    value=""
  />
</DocumentFragment>
`;
